pkg/controller/configmap_control.go

```diff
@@ -85,32 +114,48 @@ func (cc *realConfigMapControl) UpdateConfigMap(tc *v1alpha1.TidbCluster, cm *co
 
 		return updateErr
 	})
-	cc.recordConfigMapEvent("update", tc, cm, err)
+	cc.recordConfigMapEvent("update", owner, cm, err)
 	return updatedCm, err
 }
 
-func (cc *realConfigMapControl) DeleteConfigMap(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap) error {
-	err := cc.kubeCli.CoreV1().ConfigMaps(tc.Namespace).Delete(cm.Name, nil)
-	cc.recordConfigMapEvent("delete", tc, cm, err)
+func (cc *realConfigMapControl) DeleteConfigMap(owner runtime.Object, cm *corev1.ConfigMap) error {
+	err := cc.kubeCli.CoreV1().ConfigMaps(cm.Namespace).Delete(cm.Name, nil)
+	cc.recordConfigMapEvent("delete", owner, cm, err)
 	return err
 }
 
-func (cc *realConfigMapControl) recordConfigMapEvent(verb string, tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap, err error) {
-	tcName := tc.GetName()
+func (cc *realConfigMapControl) recordConfigMapEvent(verb string, owner runtime.Object, cm *corev1.ConfigMap, err error) {
+	kind := owner.GetObjectKind().GroupVersionKind().Kind
+	var name string
+	if accessor, ok := owner.(metav1.ObjectMetaAccessor); ok {
+		name = accessor.GetObjectMeta().GetName()
+	}
 	cmName := cm.GetName()
 	if err == nil {
 		reason := fmt.Sprintf("Successful%s", strings.Title(verb))
-		msg := fmt.Sprintf("%s ConfigMap %s in TidbCluster %s successful",
-			strings.ToLower(verb), cmName, tcName)
-		cc.recorder.Event(tc, corev1.EventTypeNormal, reason, msg)
+		msg := fmt.Sprintf("%s ConfigMap %s for %s/%s successful",
+			strings.ToLower(verb), cmName, kind, name)
+		cc.recorder.Event(owner, corev1.EventTypeNormal, reason, msg)
 	} else {
 		reason := fmt.Sprintf("Failed%s", strings.Title(verb))
-		msg := fmt.Sprintf("%s ConfigMap %s in TidbCluster %s failed error: %s",
-			strings.ToLower(verb), cmName, tcName, err)
-		cc.recorder.Event(tc, corev1.EventTypeWarning, reason, msg)
+		msg := fmt.Sprintf("%s ConfigMap %s for %s/%s failed error: %s",
+			strings.ToLower(verb), cmName, kind, name, err)
+		cc.recorder.Event(owner, corev1.EventTypeWarning, reason, msg)
 	}
 }
 
+// mergeConfigMap merge the desired configmap to current one
+func mergeConfigMap(current runtime.Object, desired runtime.Object) error {
+	cmC := current.(*corev1.ConfigMap)
+	cmD := desired.(*corev1.ConfigMap)
+	cmC.Data = cmD.Data
+	cmC.Labels = cmD.Labels
+	for k, v := range cmD.Annotations {
+		cmC.Annotations[k] = v
+	}
```

Comment:
why we must keep old annotations?

Reply:
It is not a must. I preserve the un-overridden annotations deliberately, which allows user and third-party components add their annotations to resources owned by TidbCluster.